walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,125 @@
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        //the time slots that mandatory attendees are unavailable
+        ArrayList<TimeRange> mandatoryAttendeesUnavailableTimes = new ArrayList<TimeRange>();
+        //the time slots that mandatory and optional attendees are unavailable
+        ArrayList<TimeRange> optionalAttendeesUnavailableTimes = new ArrayList<TimeRange>();
+        for (Event event : events) {
+            HashSet<String> intersection;
+            intersection = new HashSet<String>();
+            intersection.addAll(event.getAttendees());
+            intersection.retainAll(request.getAttendees());
+            //if a mandatory attendee has an event
+            if (intersection.size() > 0) {
+                mandatoryAttendeesUnavailableTimes.add(event.getWhen());
+                optionalAttendeesUnavailableTimes.add(event.getWhen());
+            }
+            intersection = new HashSet<String>();
+            intersection.addAll(event.getAttendees());
+            intersection.retainAll(request.getOptionalAttendees());
+            //if an optional attendee has an event
+            if (intersection.size() > 0) {
+                optionalAttendeesUnavailableTimes.add(event.getWhen());
+            }
+        }
+        //the time slots that mandatory attendees are available
+        ArrayList<TimeRange> mandatoryAttendeesAvailableTimes = findAvailableTimes(mandatoryAttendeesUnavailableTimes, request);
+        //the time slots that mandatory and optional attendees are available
+        ArrayList<TimeRange> optionalAttendeesAvailableTimes = findAvailableTimes(optionalAttendeesUnavailableTimes, request);
+        //if there are time slots that mandatory and optional attendees are available
+        if (optionalAttendeesAvailableTimes.size() > 0) {
+            return optionalAttendeesAvailableTimes;
+        }
+        else {
+            return mandatoryAttendeesAvailableTimes;
+        }
+    }
+
+    private ArrayList<TimeRange> findAvailableTimes(ArrayList<TimeRange> unavailableTimes, MeetingRequest request) {
+        //sort time ranges by start time
+        TimeRangeComparator comparator = new TimeRangeComparator();
+        unavailableTimes.sort(comparator);
+        //combine overlapping time ranges
+        int i = 0;
+        while (i < unavailableTimes.size() - 1) {
+            if (unavailableTimes.get(i).overlaps(unavailableTimes.get(i + 1)) == true) {
+                int start = unavailableTimes.get(i).start();
+                int end = (Long.compare(unavailableTimes.get(i).end(), unavailableTimes.get(i + 1).end()) > 0) ? unavailableTimes.get(i).end() : unavailableTimes.get(i + 1).end();
+                unavailableTimes.set(i, TimeRange.fromStartEnd(start, end, false));
+                unavailableTimes.remove(i + 1);
+            }
+            else {
+                i++;
+            }
+        }
+        //the time slots that attendees are available
```

Comment:
not sure if this comment is particularly informative
also, consider adding a line break here, since we're starting a new set of logic.